File: web_angular/src/app/core/services/api.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams, HttpHeaders } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { ApiResponse } from '../../shared/types/api.types';

// Detectar URL da API baseado no ambiente (igual ao api.ts original)
function getApiUrl(): string {
  // Prioridade 1: Variável de ambiente (configurada no build/deploy)
  // Angular usa NG_APP_* para variáveis de ambiente
  if (typeof (window as any).__env !== 'undefined' && (window as any).__env?.NG_APP_API_URL) {
    return (window as any).__env.NG_APP_API_URL;
  }

  // Prioridade 2: No browser, usar a mesma origem do frontend
  if (typeof window !== 'undefined') {
    const hostname = window.location.hostname;
    // Se não for localhost, usar a mesma origem do frontend
    if (hostname !== 'localhost' && hostname !== '127.0.0.1') {
      // Usar a mesma origem (protocolo + hostname) para evitar CORS
      return `${window.location.protocol}//${window.location.hostname}/api`;
    }
  }

  // Padrão para desenvolvimento local
  return 'http://localhost:8080/api';
}

@Injectable({
  providedIn: 'root'
})
export class ApiService {
  private http = inject(HttpClient);
  private apiUrl = getApiUrl();

  private getHeaders(): HttpHeaders {
    const headers: { [key: string]: string } = {
      'Content-Type': 'application/json',
    };

    // Adicionar token de autenticação se existir
    const userStr = localStorage.getItem('user');
    if (userStr) {
      try {
        const user = JSON.parse(userStr);
        if (user.token) {
          headers['Authorization'] = `Bearer ${user.token}`;
        }
        if (user.id) {
          headers['X-User-ID'] = String(user.id);
        }
      } catch (e) {
        // Ignorar erro de parsing
      }
    }

    return new HttpHeaders(headers);
  }

  private buildUrl(endpoint: string, params?: Record<string, string | number | boolean | undefined>): string {
    let url = `${this.apiUrl}${endpoint}`;
    
    if (params && Object.keys(params).length > 0) {
      let httpParams = new HttpParams();
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null && value !== '') {
          httpParams = httpParams.append(key, String(value));
        }
      });
      const queryString = httpParams.toString();
      if (queryString) {
        url += `?${queryString}`;
      }
    }

    return url;
  }

  private handleRequest<T>(request: Observable<any>): Observable<ApiResponse<T>> {
    return request.pipe(
      map((data) => {
        // Se success for explicitamente false, retornar erro
        if (data.success === false) {
          return {
            success: false,
            message: data.message || 'Erro na requisição',
          } as ApiResponse<T>;
        }
        
        // Se success for true ou não existir (assumir sucesso para códigos 2xx)
        return {
          success: data.success !== false, // true se não for false
          message: data.message,
          data: data.data || data, // Priorizar data, senão retornar tudo
          ...data,
        } as ApiResponse<T>;
      }),
      catchError((error: any) => {
        console.error('Erro na requisição:', error);
        
        // Tentar extrair mensagem do backend
        let errorMessage = 'Erro de conexão';
        
        // Angular HttpClient coloca o body do erro em error.error
        if (error.error) {
          if (typeof error.error === 'string') {
            // Se for string, tentar parsear como JSON
            try {
              const parsed = JSON.parse(error.error);
              errorMessage = parsed.message || errorMessage;
            } catch {
              errorMessage = error.error;
            }
          } else if (error.error.message) {
            errorMessage = error.error.message;
          } else if (error.error.error) {
            errorMessage = error.error.error;
          }
        } else if (error.message) {
          errorMessage = error.message;
        }
        
        return throwError(() => ({
          success: false,
          message: errorMessage,
          error: error
        } as ApiResponse<T>));
      })
    );
  }

  get<T>(endpoint: string, params?: Record<string, string | number | boolean | undefined>): Observable<ApiResponse<T>> {
    const url = this.buildUrl(endpoint, params);
    return this.handleRequest<T>(
      this.http.get<T>(url, { headers: this.getHeaders() })
    );
  }

  post<T>(endpoint: string, body: any): Observable<ApiResponse<T>> {
    const url = this.buildUrl(endpoint);
    return this.handleRequest<T>(
      this.http.post<T>(url, body, { headers: this.getHeaders() })
    );
  }

  put<T>(endpoint: string, body: any): Observable<ApiResponse<T>> {
    const url = this.buildUrl(endpoint);
    return this.handleRequest<T>(
      this.http.put<T>(url, body, { headers: this.getHeaders() })
    );
  }

  delete<T>(endpoint: string): Observable<ApiResponse<T>> {
    const url = this.buildUrl(endpoint);
    return this.handleRequest<T>(
      this.http.delete<T>(url, { headers: this.getHeaders() })
    );
  }
}

File: mobile/lib/core/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../config/api_config.dart';
import '../models/api_response.dart';

class ApiService {
  static final ApiService _instance = ApiService._internal();
  factory ApiService() => _instance;
  ApiService._internal();

  String get baseUrl => ApiConfig.baseUrl;

  // Cache para SharedPreferences
  SharedPreferences? _prefs;

  Future<SharedPreferences> _getPrefs() async {
    _prefs ??= await SharedPreferences.getInstance();
    return _prefs!;
  }

  Future<Map<String, String>> _getHeaders() async {
    final prefs = await _getPrefs();
    final headers = <String, String>{
      'Content-Type': 'application/json',
    };
    
    final token = prefs.getString('token');
    if (token != null) {
      headers['Authorization'] = 'Bearer $token';
    }
    
    final userJson = prefs.getString('user');
    if (userJson != null) {
      try {
        final userMap = jsonDecode(userJson) as Map<String, dynamic>;
        final userId = userMap['id'];
        if (userId != null) {
          headers['X-User-ID'] = userId.toString();
        }
      } catch (e) {
        // Ignorar erro de parsing
      }
    }
    
    return headers;
  }

  Future<ApiResponse<T>> get<T>(
    String endpoint, {
    Map<String, dynamic>? queryParams,
    T Function(dynamic)? fromJson,
  }) async {
    try {
      var uri = Uri.parse('$baseUrl$endpoint');
      
      if (queryParams != null && queryParams.isNotEmpty) {
        uri = uri.replace(queryParameters: queryParams.map(
          (key, value) => MapEntry(key, value.toString()),
        ));
      }

      final response = await http.get(
        uri,
        headers: await _getHeaders(),
      );

      return _handleResponse<T>(response, fromJson);
    } catch (e) {
      return ApiResponse<T>(
        success: false,
        message: 'Erro de conexão: ${e.toString()}',
      );
    }
  }

  Future<ApiResponse<T>> post<T>(
    String endpoint,
    Map<String, dynamic> body, {
    T Function(dynamic)? fromJson,
  }) async {
    try {
      final uri = Uri.parse('$baseUrl$endpoint');
      
      final response = await http.post(
        uri,
        headers: await _getHeaders(),
        body: jsonEncode(body),
      );

      return _handleResponse<T>(response, fromJson);
    } catch (e) {
      return ApiResponse<T>(
        success: false,
        message: 'Erro de conexão: ${e.toString()}',
      );
    }
  }

  Future<ApiResponse<T>> put<T>(
    String endpoint,
    Map<String, dynamic> body, {
    T Function(dynamic)? fromJson,
  }) async {
    try {
      final uri = Uri.parse('$baseUrl$endpoint');
      
      final response = await http.put(
        uri,
        headers: await _getHeaders(),
        body: jsonEncode(body),
      );

      return _handleResponse<T>(response, fromJson);
    } catch (e) {
      return ApiResponse<T>(
        success: false,
        message: 'Erro de conexão: ${e.toString()}',
      );
    }
  }

  Future<ApiResponse<T>> delete<T>(
    String endpoint, {
    T Function(dynamic)? fromJson,
  }) async {
    try {
      final uri = Uri.parse('$baseUrl$endpoint');
      
      final response = await http.delete(
        uri,
        headers: await _getHeaders(),
      );

      return _handleResponse<T>(response, fromJson);
    } catch (e) {
      return ApiResponse<T>(
        success: false,
        message: 'Erro de conexão: ${e.toString()}',
      );
    }
  }

  ApiResponse<T> _handleResponse<T>(
    http.Response response,
    T Function(dynamic)? fromJson,
  ) {
    try {
      final jsonData = jsonDecode(response.body) as Map<String, dynamic>;
      
      if (response.statusCode >= 200 && response.statusCode < 300) {
        return ApiResponse.fromJson(jsonData, fromJson);
      } else {
        return ApiResponse<T>(
          success: false,
          message: jsonData['message'] as String? ?? 
                  'Erro na requisição: ${response.statusCode}',
        );
      }
    } catch (e) {
      return ApiResponse<T>(
        success: false,
        message: 'Erro ao processar resposta: ${e.toString()}',
      );
    }
  }
}

